### PR TITLE
[amdgpu] Part5 enable the api of amdgpu

### PR DESCRIPTION
**Pull request by GaleSeLee**
Issue: https://github.com/taichi-dev/taichi/issues/6434

### Brief Summary
1. enable amdgpu api in taichi(except struct for)

### CODE REVIEW

File: taichi/runtime/llvm/llvm_runtime_executor.cpp [hunk 1]
```diff
@@ -372,7 +420,7 @@ void LlvmRuntimeExecutor::initialize_llvm_runtime_snodes(
     uint64 *result_buffer) {
   TaichiLLVMContext *tlctx = nullptr;
   if (config_->arch == Arch::cuda) {
-#if defined(TI_WITH_CUDA)
+#if defined(TI_WITH_CUDA) || defined(TI_WITH_AMDGPU)
```
Comment (jim19930609) [inline taichi/runtime/llvm/llvm_runtime_executor.cpp:423]: Put it outside of `config_->arch == Arch::cuda`

File: taichi/program/program.cpp [hunk 2]
```diff
@@ -17,6 +17,11 @@
 #include "taichi/ir/frontend_ir.h"
 #include "taichi/program/snode_expr_utils.h"
 #include "taichi/math/arithmetic.h"
+
+#if defined(TI_WITH_AMDGPU)
+#include "taichi/platform/amdgpu/detect_amdgpu.h"
```
Comment (jim19930609) [inline taichi/program/program.cpp:22]: You don't seem to use any functions from `detect_amdgpu.h` in this file?

File: taichi/program/kernel.cpp [hunk 3]
```diff
@@ -10,6 +10,10 @@
 #include "taichi/program/program.h"
 #include "taichi/util/action_recorder.h"
 
+#if defined(TI_WITH_AMDGPU)
+#include "taichi/rhi/amdgpu/amdgpu_driver.h"
```
Comment (jim19930609) [inline taichi/program/kernel.cpp:14]: Have you ever used any methods from `AMDGPUDriver` in this file?

File: taichi/codegen/llvm/codegen_llvm.cpp [hunk 4]
```diff
@@ -2639,6 +2639,12 @@ LLVMCompiledTask TaskCodeGenLLVM::run_compilation() {
       TI_ASSERT(func);
       tlctx->mark_function_as_cuda_kernel(func, task.block_dim);
     }
+  } else if (config.arch == Arch::amdgpu) {
+    for (const auto &task : offloaded_tasks) {
+      llvm::Function *func = module->getFunction(task.name);
+      TI_ASSERT(func);
+      func->setCallingConv(llvm::CallingConv::AMDGPU_KERNEL);
+    }
```
Comment (jim19930609) [inline taichi/codegen/llvm/codegen_llvm.cpp:2647]: Is it possible to organize this into a ModulePass in `llvm_context_pass.h`?

Comment (GaleSeLee) [reply taichi/codegen/llvm/codegen_llvm.cpp:2647]: Yeah. But I think this function that marks the taichi-kernel as a device-side function is simple.
What do you think about packaging it into a `mark_function_as_amdgpu_kernel` like cuda?

Comment (jim19930609) [reply taichi/codegen/llvm/codegen_llvm.cpp:2640]: Yup, I'd vote for making a `mark_function_as_amdgpu_kernel` function

**Review by jim19930609 — APPROVED** [review]

**Pull request merged**